python/cuml/cluster/hdbscan/prediction.pyx

```diff
@@ -209,6 +226,117 @@ def all_points_membership_vectors(clusterer):
+            as cpu_membership_vector
+
+        # trained on gpu
+        if not hasattr(clusterer, "_cpu_model"):
+            # the reference HDBSCAN implementations uses @property
+            # for attributes without setters available for them,
+            # so they can't be transferred from the GPU model
+            # to the CPU model
+            raise ValueError("Inferring on CPU is not supported yet when the "
+                             "model has been trained on GPU")
+
+        host_points_to_predict = input_to_host_array(points_to_predict).array
+        return cpu_membership_vector(clusterer._cpu_model,
+                                     host_points_to_predict)
+
+    elif device_type == DeviceType.device:
+        # trained on cpu
+        if hasattr(clusterer, "_cpu_model"):
+            clusterer._prep_cpu_to_gpu_prediction()
+
+    if not clusterer.fit_called_:
+        raise ValueError("The clusterer is not fit on data. "
+                         "Please call clusterer.fit first")
+
+    if not clusterer.prediction_data:
+        raise ValueError("PredictionData not generated. "
+                         "Please call clusterer.fit again with "
+                         "prediction_data=True")
+
+    points_to_predict_m, n_prediction_points, n_cols, _ = \
+        input_to_cuml_array(points_to_predict, order='C',
+                            check_dtype=[np.float32],
+                            convert_to_dtype=(np.float32
+                                              if convert_dtype
+                                              else None))
+    
+    if clusterer.n_clusters_ == 0:
+        return np.zeros(n_prediction_points, dtype=np.float32)
+
+    if n_cols != clusterer.n_cols:
+        raise ValueError('New points dimension does not match fit data!')
+
+    cdef uintptr_t prediction_ptr = points_to_predict_m.ptr
+    cdef uintptr_t input_ptr = clusterer.X_m.ptr
+    
+    membership_vec = CumlArray.empty(
+        (n_prediction_points * clusterer.n_clusters_,),
+        dtype="float32")
+
+    cdef uintptr_t membership_vec_ptr = membership_vec.ptr
+
+    cdef CondensedHierarchy[int, float] *condensed_tree = \
+        <CondensedHierarchy[int, float]*><size_t> clusterer.condensed_tree_ptr
+
+    cdef PredictionData *prediction_data_ = \
+        <PredictionData*><size_t>clusterer.prediction_data_ptr
+
+    cdef handle_t* handle_ = <handle_t*><size_t>clusterer.handle.getHandle()
+
+    compute_membership_vector(handle_[0],
+                              deref(condensed_tree),
+                              deref(prediction_data_),
+                              <float*> input_ptr,
+                              <float*> prediction_ptr,
+                              n_prediction_points,
+                              clusterer.min_samples,
+                              _metrics_mapping[clusterer.metric],
+                              <float*> membership_vec_ptr,
```

Comment:
Just to avoid any potential future issues, we should probably validate that this is actually positive (in Python) before we pass it down.